### PR TITLE
Kernelupdates

### PR DESCRIPTION
**Pull request by kmoore134**
This branch moves us to 2 stage updating again. Updates will now be done with the kernel first, then after a reboot, the rest of the update will be finished. 

Should the update fail after reboot, the previous BE will be rolled back to automatically.

### CODE REVIEW

File: update/update.go [hunk 1]
```diff
@@ -574,127 +588,49 @@ func checkBaseBootstrapSwitch() {
...
-				"install", "-U", "-f", "-y", "os/userland",
-			)
-			fullout, err := pkgcmd.CombinedOutput()
-			if err != nil {
-				ws.SendMsg(string(fullout))
-				ws.SendMsg("Failed boot-strapping userland", "fatal")
-			}
-		}
+	selfbin, _ := os.Executable()
+	ugobin := "/." + defines.ToolName
```
Comment (skarekrow) [inline update/update.go:616]: Do you mean ./?

File: update/update.go [hunk 2]
```diff
@@ -876,32 +763,81 @@ func updateincremental(force bool) error {
...
+	logger.LogToFile("Finished Kernel Update Stage 1\n-----------------------")
 
 	// Check if we need to do any ZFS automagic
 	sanitize_zfs()
 
-	// Update the bootloader
-	UpdateLoader(defines.STAGEDIR)
+}
+
+func startupgrade(kernelupdate bool) {
```
Comment (skarekrow) [inline update/update.go]: Casing

File: update/update.go [hunk 3]
```diff
@@ -911,6 +847,106 @@ func startupgrade() {
 		"Success! Reboot your system to continue the update process.",
 		"shutdown",
 	)
+
+}
+
+func preparestage2() {
```
Comment (skarekrow) [inline update/update.go]: Casing

File: update/update.go [hunk 4]
```diff
@@ -911,6 +847,106 @@ func startupgrade() {
...
+	cmd := exec.Command("mount", "-u", "rw", "/")
+	err := cmd.Run()
+	if err != nil {
+		copylogexit(err, "Failed mounting -u rw")
+	}
+
+	// Set the OLD BE as the default in case we crash and burn...
+	dat, err := ioutil.ReadFile("/.updategooldbename")
+	if err != nil {
+		copylogexit(err, "Failed read .updategooldbename")
```
Comment (skarekrow) [inline update/update.go]: Reading

File: update/update.go [hunk 5]
```diff
@@ -1005,7 +1047,12 @@ func copylogexit(perr error, text string) {
 	logger.LogToFile(perr.Error())
 	logger.LogToFile(text)
 	log.Println(text)
-	log.Fatal(perr)
+
+}
+
+// We've failed, lets reboot back into the old BE
+func rebootnow() {
```
Comment (skarekrow) [inline update/update.go]: Casing

File: update/update.go [hunk 6]
```diff
@@ -911,6 +847,106 @@ func startupgrade() {
...
+	activatebe()
+
+	// Update the bootloader
+	UpdateLoader("")
+
+	os.Exit(0)
+
+}
+
+func activatebe() {
```
Comment (skarekrow) [inline update/update.go]: Casing

**Review by skarekrow — CHANGES_REQUESTED** [review]
Just styling :D

File: update/update.go (hunk 1 above)
Comment (kmoore134) [reply update/update.go:616]: No, this was correct. Hidden binary with '/.' just for style.

**Review by skarekrow — APPROVED** [review]

**Pull request merged**